osquery/tables/system/windows/windows_events_vtable.cpp

```diff
@@ -0,0 +1,267 @@
+      std::vector<wchar_t> renderedContent;
+      unsigned long renderedBuffSize = 0;
+      unsigned long renderedBuffUsed = 0;
+      unsigned long propCount = 0;
+      EvtRender(nullptr,
+                events[i],
+                EvtRenderEventXml,
+                renderedBuffSize,
+                renderedContent.data(),
+                &renderedBuffUsed,
+                &propCount);
+
+      if (GetLastError() == ERROR_INSUFFICIENT_BUFFER) {
+        renderedBuffSize = renderedBuffUsed;
+        renderedContent.resize(renderedBuffSize);
+        EvtRender(nullptr,
+                  events[i],
+                  EvtRenderEventXml,
+                  renderedBuffSize,
+                  renderedContent.data(),
+                  &renderedBuffUsed,
+                  &propCount);
+      }
+      if (GetLastError() != ERROR_SUCCESS) {
+        LOG(WARNING) << "Failed to render windows event with "
+                     << GetLastError();
+        continue;
+      }
+
+      Row r;
+      pt::ptree propTree;
+      std::stringstream ss;
+      ss << wstringToString(renderedContent.data());
+      pt::read_xml(ss, propTree);
+      parseWelXml(propTree, r);
+
+      // Update the hidden time_range and timestamps column to avoid
+      // discarding the table silently.
+      auto time_filters = context.constraints["time_range"].getAll(EQUALS);
+      for (auto filter : time_filters) {
+        r["time_range"] = filter;
+      }
+
+      auto timestamp = context.constraints["timestamp"].getAll(EQUALS);
+      for (auto filter : timestamp) {
+        r["timestamp"] = filter;
+      }
+
+      results.push_back(r);
+      EvtClose(events[i]);
+    }
+
+    ret = EvtNext(
+        queryResults, kNumEventsBlock, events.data(), INFINITE, 0, &numEvents);
+  }
+}
+
+void genXfilterFromConstraints(QueryContext& context, std::string& xfilter) {
+  std::vector<std::string> xfilterList;
+
+  auto eids = context.constraints["eventid"].getAll(EQUALS);
+  if (!eids.empty()) {
+    xfilterList.emplace_back("(EventID=" + osquery::join(eids, " or EventID=") +
+                             ")");
+  }
+
+  auto tasks = context.constraints["task"].getAll(EQUALS);
+  if (!tasks.empty()) {
+    xfilterList.emplace_back("(Task=" + osquery::join(tasks, " or Task=") +
+                             ")");
+  }
+
+  auto levels = context.constraints["level"].getAll(EQUALS);
+  if (!levels.empty()) {
+    xfilterList.emplace_back("(Level=" + osquery::join(levels, " or Level=") +
+                             ")");
+  }
+
+  auto pids = context.constraints["pid"].getAll(EQUALS);
+  if (!pids.empty()) {
+    xfilterList.emplace_back(
+        "(Execution[@ProcessID=" +
+        osquery::join(pids, "]) or (Execution[@ProcessID=") + "])");
+  }
+
+  auto times = context.constraints["time_range"].getAll(EQUALS);
+  auto timestamps = context.constraints["timestamp"].getAll(EQUALS);
+  if (!times.empty()) {
+    for (auto t : times) {
```

Comment:
I'm assuming this `auto` is `const auto&`?

Reply:
Thanks for the catch! Using references is more appropriate here as control variable. 👍